File: android/app/src/main/java/dev/deliteai/assistant/presentation/viewmodels/HistoryViewModel.kt
/*
 * SPDX-FileCopyrightText: (C) 2025 DeliteAI Authors
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package ai.nimbleedge.nimbleedge_chatbot.presentation.viewmodels

import ai.nimbleedge.nimbleedge_chatbot.domain.models.ChatMessage
import ai.nimbleedge.nimbleedge_chatbot.domain.models.HistoryItem
import ai.nimbleedge.nimbleedge_chatbot.domain.repositories.CacheRepository
import android.app.Application
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.AndroidViewModel

class HistoryViewModel(private val application: Application) : AndroidViewModel(application) {
    private val cacheRepository = CacheRepository(application)
    var chatHistory = mutableStateOf<List<HistoryItem>?>(
        null
    )

    fun updateChatHistory(){
        chatHistory.value = retrieveChatHistory()
    }

    fun searchInChatHistory(chatId: String, searchQuery: String): Boolean {
        val chat = cacheRepository.retrieveChat(chatId)
        var searchCandidate = ""

        chat.messages.forEach {
            searchCandidate += it.message ?: ""
        }

        return searchCandidate.contains(searchQuery, ignoreCase = true)
    }

    fun deleteChat(id: String) {
        val currentHistory = chatHistory.value ?: return

        chatHistory.value = currentHistory.toMutableList().filterNot {
            it.parentChatId == id
        }

        cacheRepository.deleteChat(id)
    }

   private fun retrieveChatHistory(): List<HistoryItem> {
        val historyCards = mutableListOf<HistoryItem>()
        val chatIds = cacheRepository.getChatIds()

        chatIds.forEach {
            val chat = cacheRepository.retrieveChat(it)

            if(chat.messages.isEmpty()) return@forEach

            var firstNonEmptyChat: ChatMessage = chat.messages[0]

            for (message in chat.messages) {
                if (message.message != "") {
                    firstNonEmptyChat = message
                    break
                }
            }

            historyCards.add(
                HistoryItem(
                    chat.id,
                    firstNonEmptyChat.message ?: "Empty Conversation",
                    firstNonEmptyChat.timestamp
                )
            )
        }

        return historyCards.reversed()
    }
}

File: ios/NimbleEdgeAI/presentation/ViewModel/HistoryViewModel.swift
/*
 * SPDX-FileCopyrightText: (C) 2025 DeliteAI Authors
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation

class HistoryViewModel: ObservableObject {
    private let cacheRepository: CacheRepository
    
    @Published var chatHistory: [HistoryItem]? = nil
    @Published var chatHistoryFiltered: [ArrangedHistoryItem] = [ArrangedHistoryItem]()
    @Published var searchedChatHistory: [HistoryItem]? = nil
    
    init() {
        self.cacheRepository = CacheRepository()
    }
    
    func updateChatHistory() {
        chatHistory = retrieveChatHistory()
        chatHistoryFiltered = filterAndGroupHistory(chatHistory: chatHistory ?? [])
    }
    
    func searchInChatHistory(chatId: String, searchQuery: String) -> Bool {
        guard let chat = cacheRepository.retrieveChat(id: chatId) else{ return false}
        
        let searchCandidate = chat.messages.compactMap { $0.message }.joined()
        
        return searchCandidate.range(of: searchQuery, options: .caseInsensitive) != nil
    }
    
    private func retrieveChatHistory() -> [HistoryItem] {
        var historyCards: [HistoryItem] = []
        
        let chatIds = cacheRepository.getChatIds()
        
        for id in chatIds {
            
            guard let chat = cacheRepository.retrieveChat(id: id) else { continue }
            
            if !chat.messages.isEmpty {
                var firstNonEmptyChat = chat.messages[0]
                for message in chat.messages {
                    if let text = message.message, !text.isEmpty {
                        firstNonEmptyChat = message
                        break
                    }
                }
                
                let historyItem = HistoryItem(
                    parentChatId: chat.id,
                    title: firstNonEmptyChat.message ?? "Empty Conversation",
                    dateTime: firstNonEmptyChat.timestamp,
                    chats: chat.messages
                )
                
                historyCards.append(historyItem)
            }
            
        }
        
        return historyCards.reversed()
    }
    
    func filterAndGroupHistory(chatHistory: [HistoryItem]) -> [ArrangedHistoryItem] {
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())
        
        var todayItems: [HistoryItem] = []
        var previous3DaysItems: [HistoryItem] = []
        var previous7DaysItems: [HistoryItem] = []
        var previous30DaysItems: [HistoryItem] = []
        var olderItems: [HistoryItem] = []
        
        for item in (chatHistory ?? []) {
            let itemDate = calendar.startOfDay(for: item.dateTime)
            let diffDays = calendar.dateComponents([.day], from: itemDate, to: today).day ?? 0
            
            switch diffDays {
            case 0:
                todayItems.append(item)
            case 1...3:
                previous3DaysItems.append(item)
            case 4...7:
                previous7DaysItems.append(item)
            case 8...30:
                previous30DaysItems.append(item)
            default:
                olderItems.append(item)
            }
        }
        
        var sections: [ArrangedHistoryItem] = []
        
        if !todayItems.isEmpty {
            sections.append(ArrangedHistoryItem(section: "Today", history: todayItems))
        }
        if !previous3DaysItems.isEmpty {
            sections.append(ArrangedHistoryItem(section: "Previous 3 Days", history: previous3DaysItems))
        }
        if !previous7DaysItems.isEmpty {
            sections.append(ArrangedHistoryItem(section: "Previous 7 Days", history: previous7DaysItems))
        }
        if !previous30DaysItems.isEmpty {
            sections.append(ArrangedHistoryItem(section: "Previous 30 Days", history: previous30DaysItems))
        }
        if !olderItems.isEmpty {
            sections.append(ArrangedHistoryItem(section: "Older", history: olderItems))
        }
        
        return sections
    }
    
    func deleteChat(id: String) {
        
        cacheRepository.deleteChat(id: id)
        if let chatIDIndex = chatHistory?.firstIndex(where: { $0.parentChatId == id }) {
            chatHistory?.remove(at: chatIDIndex)
            chatHistoryFiltered = filterAndGroupHistory(chatHistory: chatHistory ?? [])
        }
        
    }
    
    func searchHistory(text: String) {
        
        if text.isEmpty {
            chatHistoryFiltered = filterAndGroupHistory(chatHistory: chatHistory ?? [])
            return
        }
        
        if let chatHistory = chatHistory {
            searchedChatHistory = chatHistory.filter { historyItem in
                historyItem.chats.contains { chat in
                    guard let message = chat.message else { return false }
                    return message.range(of: text, options: .caseInsensitive) != nil
                }
            }
            
            chatHistoryFiltered = filterAndGroupHistory(chatHistory: searchedChatHistory ?? [])
        }
    }
    
    
}
